lib/decompress.ps1

```diff
@@ -1,123 +1,121 @@
 function requires_7zip($manifest, $architecture) {
-    foreach($dlurl in @(url $manifest $architecture)) {
-        if(file_requires_7zip $dlurl) { return $true }
+    foreach ($dlurl in @(url $manifest $architecture)) {
+        if (file_requires_7zip $dlurl) {
+            return $true
+        }
     }
 }
 
 function requires_lessmsi ($manifest, $architecture) {
     $useLessMsi = get_config MSIEXTRACT_USE_LESSMSI
-    if (!$useLessMsi) { return $false }
-
+    if (!$useLessMsi) {
+        return $false
+    }
     $(url $manifest $architecture | Where-Object {
-        $_ -match '\.(msi)$'
-    } | Measure-Object | Select-Object -exp count) -gt 0
+            $_ -match '\.(msi)$'
+        } | Measure-Object | Select-Object -exp count) -gt 0
 }
 
 function file_requires_7zip($fname) {
     $fname -match '\.((gz)|(tar)|(tgz)|(lzma)|(bz)|(bz2)|(7z)|(rar)|(iso)|(xz)|(lzh)|(nupkg))$'
 }
 
 function extract_7zip($path, $to, $recurse) {
-    $output = 7z x "$path" -o"$to" -y
-    if($lastexitcode -ne 0) { abort "Exit code was $lastexitcode." }
-
-    # check for tar
-    $tar = (split-path $path -leaf) -replace '\.[^\.]*$', ''
-    if($tar -match '\.tar$') {
-        if(test-path "$to\$tar") { extract_7zip "$to\$tar" $to $true }
+    $logfile = "$(Split-Path $path)\7zip.log"
+    &(file_path 7zip 7z.exe) x "$path" -o"$to" -y | Out-File "$logfile"
+    if ($lastexitcode -ne 0) {
+        abort "Failed to extract files from $path.`nLog file:`n  $(friendly_path $logfile)"
     }
-
-    if($recurse) { Remove-Item $path } # clean up intermediate files
+    if (Test-Path $logfile) {
+        Remove-Item $logfile -Force
+    }
+    if ((strip_ext (fname $path)) -match '\.tar$' -or (fname $path) -match '\.tgz$') {
+        $listfiles = &(file_path 7zip 7z.exe) l "$path"
+        if ($lastexitcode -eq 0) {
+            $tar = ([Regex]"(\S*.tar)$").Matches($listfiles[-3]).Value
```

Comment:
Why `-3`?

Reply:
And what does this code do in overall?

Reply:
![7zip list](https://user-images.githubusercontent.com/5832170/55619450-45a61280-57cb-11e9-9d82-18e385758114.png)

This intends to get inner tar file name, it is in the third line from bottom.

Reply:
But what if there is some other file inside it? Is there some RFC / rule how these nested tars should look like?

Reply:
If the file name is `tgz` or `tar.xx`, there is only one file named `xxx.tar` in it, otherwise someone modifies it manually. 

https://www.gnu.org/software/tar/

Reply:
Mark this as resolved please.